File: src/main/dart/g0001_0100/s0098_validate_binary_search_tree/Solution.dart
// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Depth_First_Search #Tree #Binary_Tree
// #Binary_Search_Tree #Data_Structure_I_Day_14_Tree #Level_1_Day_8_Binary_Search_Tree
// #Udemy_Tree_Stack_Queue #Big_O_Time_O(N)_Space_O(log(N))
// #2024_10_06_Time_326_ms_(84.09%)_Space_151.4_MB_(31.82%)

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   int val;
 *   TreeNode? left;
 *   TreeNode? right;
 *   TreeNode([this.val = 0, this.left, this.right]);
 * }
 */
class Solution {
  bool isValidBST(TreeNode? root) {
    return _isValidBSTHelper(root, null, null);
  }

  bool _isValidBSTHelper(TreeNode? node, int? min, int? max) {
    if (node == null) {
      return true;
    }

    if ((min != null && node.val <= min) || (max != null && node.val >= max)) {
      return false;
    }

    return _isValidBSTHelper(node.left, min, node.val) &&
        _isValidBSTHelper(node.right, node.val, max);
  }
}

File: src/main/swift/g0001_0100/s0098_validate_binary_search_tree/Solution.swift
// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Depth_First_Search #Tree #Binary_Tree
// #Binary_Search_Tree #Data_Structure_I_Day_14_Tree #Level_1_Day_8_Binary_Search_Tree
// #Udemy_Tree_Stack_Queue #Top_Interview_150_Binary_Search_Tree #Big_O_Time_O(N)_Space_O(log(N))
// #2024_06_25_Time_14_ms_(93.62%)_Space_16_MB_(84.87%)

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func isValidBST(_ root: TreeNode?) -> Bool {
        return solve(root, Int.min, Int.max)
    }

    private func solve(_ root: TreeNode?, _ left: Int, _ right: Int) -> Bool {
        guard let root = root else {
            return true
        }
        if root.val <= left || root.val >= right {
            return false
        }
        return solve(root.left, left, root.val) && solve(root.right, root.val, right)
    }
}
